### DISCUSSION TITLE
Trouble setting up CORS and S3

### DISCUSSION THREAD

**Discussion by dscottboggs** [mastodon/mastodon#20664] [Q&A]
When I set my profile image, it just shows up as a black square.

The img tag is like 

```html
<img alt="" class="u-photo" src="http://files.mastodon-s3.tams.tech/accounts/avatars/109/321/100/601/072/510/original/1ba4bdbb34b6ad31.jpg" width="48" height="48">
```

but the image is visible at `https://files.mastodon-s3.tams.tech/files.mastodon.tams.tech/accounts/avatars/109/321/100/601/072/510/original/1ba4bdbb34b6ad31.jpg` -- including the bucket name `files.mastodon.tams.tech` and the protocol `https` rather than `http`. Setting `$S3_ALIAS_HOST` to include the protocol [results in a mangled url](https://github.com/mastodon/mastodon/blob/cf36ee99bb55eeecc2fd8fbc4d6b9d123b1294fa/config/initializers/content_security_policy.rb#L6).

Furthermore, the FF developer tools say 

> Content Security Policy: The page’s settings blocked the loading of a resource at http://files.mastodon-s3.tams.tech/accounts/avatars/109/321/100/601/072/510/original/1ba4bdbb34b6ad31.jpg (“img-src”).

...and indeed, my CORS header (newlines added for readability) is coming out as...

```
base-uri 'none';
default-src 'none';
frame-ancestors 'none';
font-src 'self' https://tams.tech;
img-src 'self' https: data: blob: https://tams.tech;
style-src 'self' https://tams.tech 'nonce-hF0/SuqWVtmlX3dtzYGNdg==';
media-src 'self' https: data: https://tams.tech; frame-src 'self' https:;
manifest-src 'self' https://tams.tech;
connect-src 'self' data: blob: https://tams.tech https://files.mastodon-s3.tams.tech wss://tams.tech; script-src 'self' https://tams.tech;
child-src 'self' blob: https://tams.tech;
worker-src 'self' blob: https://tams.tech
```

...which doesn't seem to set img-src to include the `$S3_ALIAS_HOST` value, except under `data: blob:`. My S3 config options from `.env.production` are:

~~~sh
S3_ENABLED=true
S3_PROTOCOL=http
S3_REGION=us-east-1
S3_ENDPOINT=http://s3:9000/
S3_HOSTNAME=http://s3:9000/
S3_BUCKET=files.mastodon.tams.tech
AWS_ACCESS_KEY_ID=[REDACTED]
AWS_SECRET_ACCESS_KEY=[REDACTED]
S3_ALIAS_HOST=files.mastodon-s3.tams.tech
~~~

## What do I need to set to get the CORS header and S3 media links to be generated correctly?

**Comment by ibigbug** [comment]
same issue here. i'd say the S3 config is a bit hard to follow and my case is some pictures are able to be loaded but some now. and sometimes a hard refresh helps and again it fails to load before another hard refresh

my setup is MinIO behind CloudFlare and I think I've setup CORS from CloudFlare 0 trust applications which is proxying my internal storeage
